tools/validate_spec.py

```diff
@@ -79,81 +78,125 @@ def cli(argv: List[str]) -> Union[int, str]:
+
+    except FileNotFoundError as e:
+        return f"specification {e}"
+
+    try:
+        validate(
+            message_value,
+            args.directory_invalid,
+            args.directory_valid,
+            args.json_output,
+            args.abort_on_error,
+        )
     except ValidationError as e:
         return f"{e}"
 
     return 0
 
 
-def validation_main(args: argparse.Namespace) -> None:
-    # pylint: disable=too-many-locals
-    def get_message_to_validate(message_path: Path, is_valid: bool) -> "OriginalMessage":
-        if not message_path.is_file():
-            raise ValidationError(f"{message_path} is not a regular file")
-        message_bytes = message_path.read_bytes()
-        message = OriginalMessage(message_bytes, is_valid, message_path.name)
-        return message
-
-    root_message_id: str = args.message_identifier
-    path_spec: Path = args.specification
-    path_invalid: Optional[Path] = args.directory_invalid
-    path_valid: Optional[Path] = args.directory_valid
-    full_output_path: Optional[Path] = args.json_output
-    abort_on_error: bool = args.abort_on_error
-    no_verification: bool = args.no_verification
-
-    path_generators: Dict[Path, Iterator[Path]] = {}
-    for path in [path_valid, path_invalid]:
-        if path is not None:
-            path_generators[path] = path.glob("*")
+def validate(
+    message_value: MessageValue,
+    directory_invalid: Optional[Path],
+    directory_valid: Optional[Path],
+    json_output: Optional[Path],
+    abort_on_error: bool,
+) -> int:
+
+    with OutputWriter(json_output) as output_writer:
+        for directory_path, is_valid_directory in [
+            (directory_valid, True),
+            (directory_invalid, False),
+        ]:
+            directory = directory_path.glob("*") if directory_path is not None else []
+            for path in directory:
+                validation_result = __validate_message(path, is_valid_directory, message_value)
+                output_writer.write_result(validation_result)
+                if not validation_result.validation_success and abort_on_error:
+                    raise ValidationError(f"aborted: message {path} was classified incorrectly")
 
+    return 0
+
+
+def __validate_message(
+    message_path: Path, valid_original_message: bool, message_value: MessageValue
+) -> "ValidationResult":
+    if not message_path.is_file():
+        raise ValidationError(f"{message_path} is not a regular file")
+    original_message = message_path.read_bytes()
+    parser_error: Optional[str] = None
+    parser_result: MessageValue = message_value.clone()
     try:
-        identifier = ID(root_message_id)
+        parser_result.parse(original_message)
+        valid_parser_result = (
+            parser_result.bytestring == original_message and parser_result.valid_message
+        )
+        if not valid_parser_result:
+            parser_error = "Invalid message"
```

Comment:
I tried to write a test to parse an incomplete (i.e. too short) message. I found no way no do that, as the parser raises an error if the bytestring of the message to be parsed is too short to get a value for the next field. Creating an incomplete message with pyrflx is only possible when setting the values by hand via the `set()` method.
This case covers too long messages. Too long messages are valid (the parser just stops parsing the last field when it got the number of required bytes) but they are obviously different from the original message.

Reply:
I see. I think the error message should still be more specific.

Reply:
Improved error message in e701f54

Reply:
> "original binary and message parsed by PyRFLX do not match"

I think that is still quite cryptic. Are you unsure if there are reasons other than too long messages?

Reply:
It is also checked that the message is `valid` and I am 99.9% sure that a message parsed by pyrflx from bytes is always valid or an error is raised during parsing. If that were the case, then the `valid` check could be removed completely. But there might be a case that I have not thought of and I see no harm in leaving the `valid` check in. Worst case would be to discover an unknown bug in the parser. Because of that, I did not write a specific error message like `message is too long`.

Reply:
As discussed separately, it would be good to have the more specific error message and ensure that all unexpected cases (potential errors in the parser) are covered by assertions.

Reply:
Done in  ab9b2b1